From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Mon, 2 Dec 2024 22:50:52 +0800
Subject: [PATCH 40/47] experiments: cpu_pstate_latencies: Add workarounds for
 Apple Fusion Architecture

On Apple A10/A10X/T2 when switching core types, the frequency may first
decrease, switch core type, and then increase again. Add workarounds to account
for such a situation.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 proxyclient/experiments/cpu_pstate_latencies.py | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/proxyclient/experiments/cpu_pstate_latencies.py b/proxyclient/experiments/cpu_pstate_latencies.py
index a36ad6a5..ffd86a8f 100755
--- a/proxyclient/experiments/cpu_pstate_latencies.py
+++ b/proxyclient/experiments/cpu_pstate_latencies.py
@@ -96,14 +96,15 @@ elif chip_id == 0x8011:
     ]
 
     MAX_PSTATE = [10]
+    LOG_ITERS = 50000
 
     def ps_to_cmd(reg, pstate):
         return 1 << 25 | (pstate & 15) << 12 | (pstate & 15) | (reg & ~0xf00f)
 
 elif chip_id == 0x8015:
     CREG = [
-        0x202e00000,
-        0x202e80000
+        0x208e00000,
+        0x208e80000
     ]
 
     MAX_PSTATE = [7, 8]
@@ -206,7 +207,11 @@ def set_pstate(cluster, pstate):
 
 print()
 
-LOG_ITERS = 10000
+if chip_id in (0x8010, 0x8011, 0x8012):
+    LOG_ITERS = 50000
+else:
+    LOG_ITERS = 10000
+
 logbuf = u.malloc(LOG_ITERS * 16)
 
 def bench_latency(cluster, cpu, from_pstate, to_pstate, verbose=False):
@@ -270,6 +275,7 @@ def bench_latency(cluster, cpu, from_pstate, to_pstate, verbose=False):
         print(f"Triggered at {tval}")
 
     thresh = 2/ (1/f_init + 1/f_end)
+    inc = f_end > f_init
 
     for i in range(tidx, LOG_ITERS - window - 1):
         ts0, cyc0 = log[i - window]
